---
title: "DB create and load"
output: html_document
---

## Setup

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(dbplyr)
```

Make sure database file exists

```{bash}
touch ../data/gs_intensities.sqlite
```

Load the database and connection info
```{r}
library(DBI)
library(odbc)
library(config)
dsn <- get("gsint")
con <- dbConnect(odbc::odbc(),driver = dsn$driver, database = dsn$database , timeout = 100)
```

## Setup the database

The main design of the database is 3 tables: the main data table, 

### Create tables


Create the table for the markers
```{sql, connection = con}
CREATE TABLE IF NOT EXISTS "marker_info" (
  "name" TEXT NOT NULL,
  "address" INTEGER,
  "chr" CHAR NOT NULL,
  "position" INTEGER NOT NULL,
  primary key (chr, position, name)
);

```

create batch table
```{sql, connection = con}
CREATE TABLE IF NOT EXISTS "batch"(
  "batchid" INTEGER NOT NULL,
  "chip" CHAR NOT NULL,
  primary key (batchid)
)
;

```

populate the batch info
```{sql, connection = con}
INSERT INTO batch (batchid, chip) VALUES(1,'v1.0'), (2,'v1.0');
```

create the table to store the info about the samples
```{sql, connection = con}
CREATE TABLE IF NOT EXISTS "sample"(
  "sampleid" INTEGER PRIMARY KEY AUTOINCREMENT,
  "samplecode" CHAR NOT NULL,
  "batchid" INTEGER NOT NULL,
  "barcode_pos" CHAR,
  "reported_sex" CHAR,
  "genetic_sex" CHAR,
  "ancestry" CHAR,
  "callrate" REAL,
  "passed_gt_qc" LOGICAL,
  UNIQUE(samplecode, batchid),
  FOREIGN KEY ("batchid") REFERENCES batch("batchid")
)
;
```




create the main data table
```{sql, connection = con}
CREATE TABLE IF NOT EXISTS "intensities"(
  "markername" CHAR NOT NULL,
  "sampleid" CHAR NOT NULL,
  "gtype" CHAR,
  "x_raw" REAL,
  "x" REAL,
  "y_raw" REAL,
  "y" REAL,
  'r' REAL,
  'theta' REAL,
  UNIQUE(markername,sampleid),
  FOREIGN KEY ("markername") REFERENCES marker_info("name"),
  FOREIGN KEY ("sampleid") REFERENCES sample("sampleid")
);
```

```{sql, connection = con}
CREATE VIEW IF NOT EXISTS combined AS SELECT marker_info.name, marker_info.chr, marker_info.position, sample.samplecode, sample.reported_sex, sample.genetic_sex, sample.passed_gt_qc, sample.ancestry, batch.*, intensities.* FROM intensities 
  LEFT JOIN marker_info on intensities.markername = marker_info.name 
  LEFT JOIN sample on sample.sampleid = intensities.sampleid
  LEFT JOIN batch on sample.batchid = batch.batchid;
```

## Data load

Create variables needed to filter and wrangle data
```{r}
# columns to extract for intensities table
suffixes <- c("GType", "X_Raw","X","Y_Raw","Y", "R", "Theta")
names(suffixes) <- suffixes

# possible values for chromosomes in the markers_info table
possible_chrs <- c("0","1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12", "13", "14", "15", "16", "17", "18", "19", "20", "21", "22", "MT", "X", "XY", "Y" )

intensities_colnames <- dbListFields(con, "intensities")

qc_samples <- read_csv(here::here("data/Summary_QC_PassFail101116.csv"), col_types = "ccdcccccccci___ccccc") %>% 
  rename('samplecode' = Sample.ID,  "reported_sex" = Reported.Sex, "genetic_sex" = Genetic.Sex) %>% # closest match to the original GS name
  mutate(passed_gt_qc = if_else(QCStatus == "Passed Genotyping QC", TRUE, FALSE, FALSE), 
         barcode_pos = paste0(ChipBarcode, "_",ChipPosition),
         ancestry = "none_assigned", # remove this once real ancestry is in the file
         batchid = str_extract_all(QCBatch, '[0-9]+'), rn = row_number())%>% unnest() %>% mutate(batchid = as.numeric(batchid)) %>% select(samplecode, batchid, barcode_pos, reported_sex, genetic_sex, callrate = Call.Rate, passed_gt_qc )


```

function to process the data as it comes in
```{r}

def_col_data_type <-function(filename){
  gs_header <- read_delim(file = filename, delim = '\t', n_max = 100) %>% slice(0)
  gs_header_spec <- spec(gs_header)
  gs_header_spec$cols$Chr <- col_character()
  gs_header_spec
}

# example
# def_col_data_type(here::here("data/test10000_batch1.txt"))

update_marker_table <- function(gs_tab){
  existing_markers <- dbReadTable(con, name = "marker_info")
  marker_info <- gs_tab %>% filter(Chr %in% possible_chrs) %>% select(name = Name, address = Address, chr = Chr, position = Position) %>% mutate_if(is.numeric, as.integer)
  DBI::dbWriteTable(con, name = "marker_info", value = anti_join(marker_info, existing_markers, by = c("name","address","chr","position") ) , append = TRUE)
}

update_sample_table <- function(gs_long_combined, batch){
  # existing samples
  existing_samples <- dbReadTable(con, "sample")
  
  samples <- gs_long_combined %>% select(samplecode = sample) %>% mutate(batchid = batch) %>% distinct()
  samples <- left_join(samples, qc_samples, by = c("samplecode", "batchid"))
 
  # create table of new samples to be added
  if(NROW(existing_samples) > 0){
   samples_new <- samples %>% anti_join(., existing_samples, by = c("samplecode", "batchid"))
  } else {
    samples_new <- samples
  }
  if(NROW(samples_new)> 0){
  # add new samples to db
  DBI::dbWriteTable(con, name = "sample", value = samples_new, append = TRUE)
  }
  
}

process_chunk <- function(gs_tab, batch){
  gs_tab <- gs_tab %>% set_names(., nm = str_replace(colnames(.), " ", "_")) %>% select(Index, Name, Address, Chr, Position, GenTrain_Score, starts_with("Frac_"), ends_with("GType"), ends_with(".X_Raw"), ends_with(".X"), ends_with(".Y_Raw"), ends_with("Y"), ends_with(".R"), ends_with(".Theta")) 
  # extract and write markers into db
  update_marker_table(gs_tab)
  message("markers updated")
  # make sure only adding new samples
  # make into long format
  gs_long <- purrr::map(suffixes, ~ gs_tab %>% select(Name, ends_with(paste0(".",.x))) %>% gather("sample", !!.x , ends_with(paste0(".", .x))) %>% mutate(sample = str_remove(sample, paste0("\\.",.x ))) )
  #join the results back together
  gs_long_combined <- purrr::reduce(gs_long, left_join, by = c("Name", "sample"))
  message("in long format")
  update_sample_table(gs_long_combined, batch)
  message("samples updated")
   
  gs_long_combined %>% 
    mutate(batchid = batch) %>% 
    left_join(., tbl(con, "sample") %>% select(samplecode, batchid, sampleid) ,
              by = c("sample" = "samplecode", "batchid"), 
              copy = TRUE) %>% 
    select(markername = Name, 
           sampleid = sampleid, 
           gtype = GType, 
           x_raw = X_Raw, 
           x = X, 
           y_raw = Y_Raw, 
           y = Y, 
           r = R, 
           theta = Theta) %>% 
    arrange(markername) %>% 
    select(!!dbListFields(con, "intensities")) %>% 
    write_csv(path = here::here(paste0("data/intensities_batch_",batch,".csv")), 
              col_names = FALSE, 
              append = TRUE)
  message("intensities chunk written")
}


```

```{r}
f <- function(gs_tab, pos){
  process_chunk(gs_tab, batch = 1)
}

header_spec <- def_col_data_type(here::here("data/QC_Batch1-GenomeStudio_FullDataTable.txt"))

#test <- read_delim(file = here::here("data/test10000_batch1.txt"),delim = '\t', col_types = header_spec$cols,  n_max = 100) 

read_delim_chunked(file = here::here("data/QC_Batch1-GenomeStudio_FullDataTable.txt"),delim = '\t', col_types = header_spec$cols, chunk_size = 5000, callback = SideEffectChunkCallback$new(f), progress = TRUE) 

f <- function(gs_tab, pos){
  process_chunk(gs_tab, batch = 2)
}

header_spec <- def_col_data_type(here::here("data/QC_Batch2-GenomeStudio_FullDataTable.txt"))

#test <- read_delim(file = here::here("data/test10000_batch1.txt"),delim = '\t', col_types = header_spec$cols,  n_max = 100) 

read_delim_chunked(file = here::here("data/QC_Batch2-GenomeStudio_FullDataTable.txt"),delim = '\t', col_types = header_spec$cols, chunk_size = 1000, callback = SideEffectChunkCallback$new(f), progress = TRUE)
```

Load in the intensities:
```
sqlite3 gs_intensities.sqlite

.mode csv
.import intensities_batch_1.csv intensities
```

Load in the first 100 rows to get the data types
```{r}
gs_header <- read_delim(file = here::here("data/test10000_batch1.txt"), delim = '\t', n_max = 100) %>% slice(0)
gs_header_spec <- spec(gs_header)
gs_header_spec$cols$Chr <- col_character()
```

```{r}
gs_tab <- read_delim(file = here::here("data/test10000_batch1.txt"),delim = '\t', col_types = gs_header_spec$cols) %>% set_names(., nm = str_replace(colnames(.), " ", "_")) %>% select(Index, Name, Address, Chr, Position, GenTrain_Score, starts_with("Frac_"), ends_with("GType"), ends_with(".X_Raw"), ends_with(".X"), ends_with(".Y_Raw"), ends_with("Y"), ends_with(".R"), ends_with(".Theta"))
```


```{r}

marker_info <- gs_tab %>% filter(Chr %in% possible_chrs) %>% select(markerid = Index,name = Name, address = Address, chr = Chr, position = Position) %>% mutate_if(is.numeric, as.integer)
```

```{r}
DBI::dbWriteTable(con, name = "marker_info", value = marker_info, append = TRUE)
```

check markers were inserted
```{sql, connection = con}
SELECT * from marker_info limit 5;
```

make the samples into long format for the db




```{r}
gs_long <- purrr::map(suffixes, ~ gs_tab %>% select(Index, ends_with(paste0(".",.x))) %>% gather("sample", !!.x , ends_with(paste0(".", .x))) %>% mutate(sample = str_remove(sample, paste0("\\.",.x ))) )

```


join the results back together
```{r}
gs_long_combined <- purrr::reduce(gs_long, left_join, by = c("Index", "sample"))

```

populate the batch table with the sample info
```{r}
samples <- gs_long_combined %>% select(samplecode = sample) %>% mutate(batchid = 1, sex = NA, ancestry = NA) %>% distinct()
DBI::dbWriteTable(con, name = "sample", value = samples, append = TRUE)
```


```{r}
samples_ids <- samples %>% left_join(., tbl(con, "sample") %>% select(samplecode, batchid, sampleid), by = c("samplecode", "batchid"), copy = TRUE)  %>% select(samplecode, sampleid)
```

```{r}
gs_long_combined %>% left_join(., samples_ids, by = c("sample" = "samplecode"))%>% select(markerid = Index, sampleid = sampleid, gtype = GType, x_raw = X_Raw, x = X, y_raw = Y_Raw, y = Y, r = R, theta = Theta) %>% arrange(markerid) %>% select(!!dbListFields(con, "intensities")) %>% write_csv(path = here::here("data/intensities_batch_1.csv"), col_names = FALSE)
```


```{r, eval = FALSE}
DBI::dbWriteTable(con, name = "intensities", value = gs_long_combined %>% left_join(., samples_ids, by = c("sample" = "samplecode"))%>% select(markerid = Index, sampleid = sampleid, gtype = GType, x_raw = X_Raw, x = X, y_raw = Y_Raw, y = Y, r = R, theta = Theta) %>% arrange(markerid), append = TRUE)
```



```{r}
dbDisconnect(con)
```

